py-polars/polars/utils/udfs.py

```diff
@@ -222,14 +208,54 @@ def to_expression(self, col: str) -> str | None:
                 param_name=self._param_name,
                 depth=int(bool(logical_instructions)),
             )
-            for offset, ops in logical_instruction_blocks.items()
+            for offset, ops in control_flow_blocks.items()
         }
 
-        for inst in logical_instructions:
-            expression_strings[inst.offset] = InstructionTranslator.op(inst)
-
-        # TODO: handle mixed 'and'/'or' blocks (e.g. `x > 0 AND (x > 0 OR (x == -1)`).
-        #  (need to reconstruct the correct nesting boundaries to do this properly)
+        if logical_instructions:
+            # reconstruct nesting boundaries for mixed and/or ops by associating
+            # control flow jump offsets with their target expression blocks and
+            # injecting appropriate parentheses
+            combined_offset_idxs = set()
+            if len({inst.opname for inst in logical_instructions}) > 1:
+                block_offsets: list[int] = list(expression_strings.keys())
+                idx_min, idx_max = 0, len(block_offsets) - 1
+                previous_logical_opname = ""
+                for i, inst in enumerate(logical_instructions):
+                    # operator precedence means that we can combine logically connected
+                    # 'and' blocks into one (depending on follow-on logic) and should
+                    # parenthesise nested 'or' blocks
+                    logical_op = OpNames.CONTROL_FLOW[inst.opname]
+                    start = block_offsets[
+                        max(idx_min, bisect_left(block_offsets, inst.offset) - 1)
```

Comment:
why is this `max` necessary?

is there any test where `idx_min > bisect_left(block_offsets, inst.offset) - 1`?

Reply:
There used to be - though might not be the case now, after I clarified the logic; will poke at it a bit and see if we can take it out now 🤔

Reply:
> is there any test where `idx_min > bisect_left(block_offsets, inst.offset) - 1`?

Great catch - looks like the cleaner logic means we can indeed ditch it now; given how tricky this particular block was to get right I am wildly in favour of additional simplifications... Just updated accordingly (moved the nesting code into its own function while I was at it too).